### PR TITLE
[Pass] FuseTIR

### PR DESCRIPTION
**Pull request by jinhongyii**
This pass relies on RegenerateDef which hasn't been merged, so the changed files will include its relevant files. Please skip them when reviewing.

Tuple-related tests rely on https://github.com/tlc-pack/relax/pull/116. After it is merged, I will update the tests.

### CODE REVIEW

**Comment by jinhongyii** [comment]
cc: @MasterJH5574 @Hzfengsy @tqchen

File: src/relax/transform/fuse_tir.cc [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+    static const Op& call_tir_op_ = Op::Get("relax.call_tir");
+    Expr e = ExprMutator::VisitExpr_(call);
+    call = e.as<CallNode>();
+    ICHECK(call != nullptr);
+    if (call->op->IsInstance<GlobalVarNode>()) {
+      // Handle subfunction
+      GlobalVar gv = Downcast<GlobalVar>(call->op);
+      BaseFunc func = mod_->Lookup(gv);
+      func_info_ = {};
+      in_subfunction_ = true;
```
Comment (Hzfengsy) [inline src/relax/transform/fuse_tir.cc:290]: I guess `primitive_func` may be a better name

File: src/relax/transform/fuse_tir.cc [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+      GlobalVar gv = Downcast<GlobalVar>(call->args[0]);
+      tir::PrimFunc func = Downcast<tir::PrimFunc>(mod_->Lookup(gv));
+      GlobalVar new_gv = this->builder_->AddFuncToContext(func, gv->name_hint);
+      return Call(call->op, {new_gv, call->args[1], call->args[2]}, call->attrs, call->type_args,
+                  call->span);
+    } else {
+      // Handle call_tir in subfunction
+      GlobalVar gv = Downcast<GlobalVar>(call->args[0]);
+      tir::PrimFunc func = Downcast<tir::PrimFunc>(mod_->Lookup(gv));
+      const Array<Expr> call_tir_args = Downcast<Tuple>(call->args[1])->fields;
```
Comment (Hzfengsy) [inline src/relax/transform/fuse_tir.cc:311]:
```suggestion
      const Array<Expr>& call_tir_args = Downcast<Tuple>(call->args[1])->fields;
```

File: src/relax/transform/fuse_tir.cc [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+  /*! \brief The IRModule */
+  const IRModule& mod_;
+  /*! \brief The helper info to fuse TIR primfunc */
+  FuseFuncInfo func_info_;
+  /*! \brief Whether the mutator is handling the subfunction*/
+  bool in_subfunction_;
+};
+
+/*! \brief Inline a subfunction into main*/
+class Inliner : public ExprMutator {
```
Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:482]: inline is a very generic name, consider a more specific name FuseTIRInliner

Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:482]: inline is a very generic name, consider a more specific name FuseTIRInliner

File: src/relax/transform/fuse_tir.cc [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+      } else {
+        ICHECK(false) << "Unsupported binding";
+      }
+    }
+    // Update return value remap
+    Var return_var = Downcast<Var>(seq->body);
+    this->var_remap_[binding_var->vid] = this->var_remap_[return_var->vid];
+  }
+
+  Optional<Function> get_relax_call(const VarBindingNode* binding) {
```
Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:569]: CamelCase

Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:569]: CamelCase

File: src/relax/transform/fuse_tir.cc [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+    Expr value = this->VisitExpr(binding->value);
+    if (const auto* call = value.as<CallNode>()) {
+      bool is_dataflow_var = binding->var->IsInstance<DataflowVarNode>();
+      if (call->op.same_as(call_tir_op_)) {
+        Optional<tir::PrimFunc> opt_f = MatchPrimFunc(call->args[0]);
+        if (!opt_f) {
+          LOG(FATAL) << "Cannot find the prim_func of the call_tir in the module: "
+                     << call->args[0];
+        }
+        tir::PrimFunc func = opt_f.value();
```
Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:421]: Logic flow: FusePrimFunc into a separate sub function

Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:421]: Logic flow: FusePrimFunc into a separate sub function

File: src/relax/transform/fuse_tir.cc [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+      // This is the output buffer of the fused function
+      // Add it to param and buffer map
+      for (int i = n - output_size; i < n; i++) {
+        func_info_.params.push_back(func->params[i]);
+        func_info_.buffer_map.Set(func->params[i], func->buffer_map.at(func->params[i]));
+      }
+    }
+  }
+
+  void VisitBinding_(const VarBindingNode* binding) final {
```
Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:407]: Likely you can and should overload `VisitExpr_(const CallNode*)` instead

Comment (jinhongyii) [reply src/relax/transform/fuse_tir.cc:407]: I have already overloaded VisitExpr_(const CallNode*)

File: src/relax/transform/fuse_tir.cc [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+    static const Op& call_tir_op_ = Op::Get("relax.call_tir");
+    if (!in_subfunction_) {
+      return ExprMutator::VisitBinding_(binding);
```
Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:410]: Construct a separate visitor(not mutator) TIRFuncFusor to collect the information needed for fusion

**Review by tqchen — CHANGES_REQUESTED** [review]

**Comment by jinhongyii** [comment]
@tqchen @Hzfengsy @MasterJH5574  I've refactored FuseTIR and you can have another round of review

File: src/relax/transform/fuse_tir.cc [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+      LOG(FATAL) << "Unsupported binding value: " << binding->value;
+    }
+  }
+
+  /*! \brief The IRModule */
+  const IRModule& mod_;
+  /*! \brief The helper info to fuse TIR primfunc */
+  FuseFuncInfo func_info_;
+  /*! \brief The tir function after fusion*/
+  tir::PrimFunc fused_tir;
```
Comment (tqchen) [inline src/relax/transform/fuse_tir.cc:563]: STYLE: this is something that we must always follow in instinct and avoid code review effort. Private member should all have an trailing underscore.

**Pull request merged**